src/Form/AppSettingsForm.php

```diff
@@ -69,90 +92,70 @@ public function buildForm(array $form, FormStateInterface $form_state) {
+        'progress' => [
+          'type' => 'throbber',
+          'message' => '',
         ],
       ],
     ];
 
-    /** @var string[] $default_products */
-    $default_products = $generalConfig->get('default_products') ?: [];
-    $product_list = [];
-    /** @var \Drupal\apigee_edge\Entity\ApiProduct[] $products */
-    try {
-      $products = ApiProduct::loadMultiple();
-      foreach ($products as $product) {
-        $product_list[$product->id()] = $product->getDisplayName();
-      }
-    }
-    catch (EntityStorageException $e) {
-      // Apigee Edge credentials are missing/incorrect or something else went
-      // wrong. Do not redirect the user to the error page.
-      $product_list = [];
-      $this->messenger()->addError($this->t('Unable to retrieve API product list from Apigee Edge. Please ensure that <a href=":link">Apigee Edge connection settings</a> are correct.'), [
-        ':link' => Url::fromRoute('apigee_edge.settings')->toString(),
-      ]);
-    }
+    // It's necessary to add a wrapper, because if the ID is added to the
+    // checkboxes form element, then that will not be properly rendered
+    // (the label is duplicated).
+    $form['api_product']['default_api_product_multiple_container'] = [
+      '#type' => 'container',
+      '#id' => 'default-api-product-multiple',
+    ];
 
-    $form['api_product']['default_api_product_multiple'] = [
+    $form['api_product']['default_api_product_multiple_container']['default_api_product_multiple'] = [
       '#type' => 'checkboxes',
       '#title' => $this->t('Default API Product'),
       '#options' => $product_list,
       '#default_value' => $default_products,
-      '#states' => [
-        'visible' => [
-          [
-            ':input[name="user_select"]' => [
-              'checked' => TRUE,
-              'visible' => TRUE,
-            ],
-          ],
-        ],
-      ],
+      '#required' => $form_state->getValue('user_select') === NULL ? !(bool) $generalConfig->get('user_select') : !(bool) $form_state->getValue('user_select'),
     ];
 
-    return parent::buildForm($form, $form_state);
+    return $form;
+  }
+
+  /**
+   * Ajax callback for the "user_select" checkbox.
+   *
+   * Set 'default_api_product_multiple' checkboxes form element as required
+   * if the 'user_select' is unchecked, else not required.
+   * Use AJAX instead of #states because #required in #states is not
+   * working properly.
+   *
+   * @param array $form
+   *   An associative array containing the structure of the form.
+   * @param \Drupal\Core\Form\FormStateInterface $form_state
+   *   The current state of the form.
+   *
+   * @return array
+   *   The renderable array.
+   *
+   * @see https://www.drupal.org/project/drupal/issues/2855139
+   */
+  public function apiProductListCallback(array &$form, FormStateInterface $form_state) : array {
+    return $form['api_product']['default_api_product_multiple_container'];
```

Comment:
I prefer if AJAX commands gets returned, like [ReplaceCommand](https://api.drupal.org/api/drupal/core%21lib%21Drupal%21Core%21Ajax%21ReplaceCommand.php/class/ReplaceCommand/8.5.x), instead of render elements directly.

https://api.drupal.org/api/drupal/core%21core.api.php/group/ajax/8.5.x